### DISCUSSION TITLE
how to add projection operators as perturbation?

### DISCUSSION THREAD

**Discussion by san-dbox** [QuSpin/QuSpin#505] [Q&A]
Hi, I want to add some projection operators as perturbation. It projects onto the subspace whose elements have some specific structure such as
<img src="https://latex.codecogs.com/png.image?\dpi{110}&space;\bg_white&space;H&space;=&space;\sum_j&space;|\uparrow\rangle\langle\uparrow|_j" title="\bg_white H = \sum_j |\uparrow\rangle\langle\uparrow|_j," />
or
<img src="https://latex.codecogs.com/png.image?\dpi{110}&space;\bg_white&space;H&space;=&space;\sum_j|\psi\rangle\langle\psi|_{j,&space;j&plus;1},&space;\quad\mathrm{where}\quad&space;|\psi\rangle_{j,&space;j&plus;1}&space;=&space;(|\uparrow\downarrow\rangle-|\downarrow\uparrow\rangle)_{j,&space;j&plus;1}." title="\bg_white H = \sum_j|\psi\rangle\langle\psi|_{j, j+1}, \quad\mathrm{where}\quad |\psi\rangle_{j, j+1} = (|\uparrow\downarrow\rangle-|\downarrow\uparrow\rangle)_{j, j+1}." />

Is there any simple way to describe these projection operators? I write them down as the combination of multiple spin operators. Still, this method makes the code too complex in some cases.

**Comment by mgbukov** [comment]
I'm not aware of any specific way to code in projectors; in a sense projectors are not physical operators (they are nonlocal), whence the increased complexity. 

That said, it is possible to feed a matrix directly into the hamiltonian constructor though, and turn it into a `hamiltonian` object, see definition of the arguments of [`static_list`](https://weinbe58.github.io/QuSpin/generated/quspin.operators.hamiltonian.html#quspin.operators.hamiltonian). Such a matrix can be the projector constructed out of some states.

**Reply by san-dbox** [reply]
Thanks! I missed that `static_list` can contain matrices. I'm going to try it. Where can I see an example using a matrix as a parameter of `static_list`?
